common/flatpak-dir.c

```diff
@@ -2393,6 +2393,281 @@ flatpak_dir_find_latest_rev (FlatpakDir               *self,
+{
+  int i;
+
+  for (i = 0; results[i] != NULL; i++)
+    g_hash_table_remove (results[i]->ref_to_checksum,cr);
+}
+
+typedef struct {
+  FlatpakDirResolve *resolve; /* not owned */
+  OstreeCollectionRef collection_ref; /* owns the collection_id member only */
+  char *local_commit;
+} FlatpakDirResolveData;
+
+static void
+flatpak_dir_resolve_data_free (FlatpakDirResolveData *data)
+{
+  if (data == NULL)
+    return;
+
+  g_free (data->collection_ref.collection_id);
+  g_free (data->local_commit);
+  g_free (data);
+}
+
+static gboolean
+flatpak_dir_do_resolve_p2p_refs (FlatpakDir               *self,
+                                 FlatpakDirResolveData   **datas,
+                                 gboolean                  with_commit_ids,
+                                 GCancellable             *cancellable,
+                                 GError                  **error)
+{
+  g_autoptr(GPtrArray) collection_refs_to_fetch = g_ptr_array_new ();
+  g_autoptr(GPtrArray) commit_ids_to_fetch = NULL;
+  g_autoptr(GAsyncResult) find_result = NULL;
+  g_auto(OstreeRepoFinderResultv) results = NULL;
+  g_autoptr(GVariant) find_options = NULL;
+  g_auto(GVariantBuilder) find_builder = FLATPAK_VARIANT_BUILDER_INITIALIZER;
+  GVariantBuilder pull_builder;
+  g_autoptr(GVariant) pull_options = NULL;
+  g_autoptr(GAsyncResult) pull_result = NULL;
+  g_autoptr(GMainContextPopDefault) main_context = NULL;
+  g_autoptr(FlatpakRepoTransaction) transaction = NULL;
+  g_autoptr(OstreeRepo) child_repo = NULL;
+  g_auto(GLnxLockFile) child_repo_lock = { 0, };
+  g_autoptr(GFile) user_cache_dir = NULL;
+  g_autoptr(FlatpakTempDir) child_repo_tmp_dir = NULL;
+  int i;
+
+  main_context = flatpak_main_context_new_default ();
+
+  if (with_commit_ids)
+    commit_ids_to_fetch = g_ptr_array_new ();
+
+  for (i = 0; datas[i] != NULL; i++)
+    {
+      FlatpakDirResolveData *data = datas[i];
+      FlatpakDirResolve *resolve = data->resolve;
+
+      g_ptr_array_add (collection_refs_to_fetch, &data->collection_ref);
+      if (commit_ids_to_fetch)
+        {
+          g_assert (resolve->opt_commit != NULL);
+          g_ptr_array_add (commit_ids_to_fetch, resolve->opt_commit);
+        }
+   }
+
+  g_ptr_array_add (collection_refs_to_fetch, NULL);
+
+  g_variant_builder_init (&find_builder, G_VARIANT_TYPE ("a{sv}"));
+  if (commit_ids_to_fetch)
+    g_variant_builder_add (&find_builder, "{s@v}", "override-commit-ids",
+                           g_variant_new_variant (g_variant_new_strv ((const char * const*)commit_ids_to_fetch->pdata,
+                                                                      commit_ids_to_fetch->len)));
+  find_options = g_variant_ref_sink (g_variant_builder_end (&find_builder));
+
+  /* We create a termporary child repo in the user homedir so that we can just blow it away when we're done.
+   * This lets us always write to the directory in the system-helper case, but also lets us properly clean up
```

Comment:
Don't you also have write access to the child repo in the system-helper case? I don't understand the reason for unconditionally using the user installation for this. Don't we want to have access to commits in the system installation?

Reply:
Oh, we don't use the *user installation*. We make a child repo of whichever installation we're working on, just we always store the temporary files in the users homedir, rather than in the location we use for child repos that will eventually be commited back to the system installation.